source/escalonador.c

```diff
@@ -112,16 +278,105 @@ const char * parse_clarg_topology( int argc, char *argv[] )
+              message_received.delta_delay);
+
+        // now we add the new entry to the execution queue
+        if( ( exec_queue->len > 0 ) && ( (previous_timer = alarm( TIME_OUT )) > 0 ) )
+        {
+          /*
+              if the execution_queue is not empty and there is a timer in countdown
+          */
+
+          if( previous_timer > message_received.delta_delay )
+          {
+              /*
+                check if the delay for the new entry is smaller than the current one (already in countdown)
+                if so,   i) set the new entry as the new head;
+                          ii) set the remaining delay of the previous_timer;
+                          iii) fire the alarm with the delay of the new head;
+                !!! we will always assume the execution queue is sorted (MAKE SURE IT IS) !!!
+              */
+              exec_queue->head->delay = previous_timer;
+
+              execution_entry_t *new_entry = createEntry(message_received.filename,
+                                                        message_received.delta_delay);    
+              if ( addEntry(new_entry, exec_queue) < 0 )
+              {
+                printf("LINKED_LIST_ERROR: Adding new entry to linked-list failed\n");
+              }
+
+              alarm(message_received.delta_delay);
+          }
+
+          else
+          {
+              /* 
+                if the delay for the new entry is bigger or equal to the current one (already in countdown)
+                    i) update time left of the linked list's first entry;
+                    ii) refire the previous_timer;
+                    iii) add the new entry into place (SORTED!).
+              */
+              exec_queue->head->delay = previous_timer;
+              alarm(previous_timer);
+
+              execution_entry_t *new_entry = createEntry(message_received.filename,
+                                                        message_received.delta_delay);
+              if( addEntry(new_entry, exec_queue) < 0 )
+              {
+                printf("LINKED_LIST_ERROR: Adding new entry to linked-list failed\n");
+              }
+          } 
+        }
+
+        else if ( exec_queue->len == 0 )
+        {
+          /*
+              no other process is waiting in the execution_queue
+              add the new entry
+          */
+            execution_entry_t *new_entry = createEntry(message_received.filename,
+                                                      message_received.delta_delay);    
+            if ( addEntry(new_entry, exec_queue) < 0 )
+            {
+              printf("LINKED_LIST_ERROR: Adding new entry to linked-list failed\n");
+            }
+
+            alarm(message_received.delta_delay);
+        }
+
+        else
+        {
+          /* !! if this part is reached, it's likely that there has been some sort of unwanted behavior !! */
+        }
+      }
+
+      if ( flag == 1 )
```

Comment:
Could you explain what this last block does, please? It seems the flag is set in the alarm_handler, but it feels like removing the node that was pushed into execution should also be removed in the alarm_handler. Is there a particular reason for its being here in the main()?

Reply:
I think it is easy to do it here because if not, we would have to pass the exec_queue as an argument to alarm_handler (I do not know if that is possible) or make it a global variable to be accessed in alarm_handler. I can change that in the next commit if we all agree.